tracing-subscriber/src/registry/sharded.rs

```diff
@@ -235,7 +235,11 @@ impl Subscriber for Registry {
     fn clone_span(&self, id: &span::Id) -> span::Id {
         let span = self
             .get(&id)
-            .unwrap_or_else(|| panic!("tried to clone {:?}, but no span exists with that ID", id));
+            .unwrap_or_else(|| panic!(
+                "tried to clone {:?}, but no span exists with that ID\n\
+                This may be caused by consuming a parent span (`parent: span`) rather than borrowing it (`parent: &span`).",
+                id,
```

Comment:
Another case where this might occur is if a span is cloned by a subscriber other than the one that originally assigned its ID. 

This will never happen when using the `tracing::Span` type (barring implementation bugs), since the `Span` retains a reference to the subscriber that assigned its ID, and all operations performed by that span will be performed by that subscriber. However, it may happen in code where `Subscriber::clone_span()` is manually called on a `tracing_core::Id`, if that ID was provided by a different subscriber. I wonder if it's also worth mentioning this in the panic message?

Reply:
I'd be happy to expand the message further if you think that case is common enough to provide debugging assistance for.

I also wonder if a future (incompatible) version of tracing could somehow tag the `Id` based on the subscriber that issued it, so that you *can't* mix `Id` values from different subscribers. (They could all implement a common trait and the type difference could be erased at runtime, but the concrete type could somehow reference the Subscriber using a lifetime.)

Reply:
> I also wonder if a future (incompatible) version of tracing could somehow tag the `Id` based on the subscriber that issued it, so that you _can't_ mix `Id` values from different subscribers.

This is essentially what the `tracing::Span` type represents; it's an `Id` plus a runtime reference to a subscriber. I'm not sure if tagging IDs with subscriber _types_ at compile-time makes a lot of sense — the loose coupling between the instrumentation-side API with the subscriber is kind of central to how `tracing` provides a facade layer allowing the subscriber to be modular. Something like what you're proposing here could maybe be possible, but it would only be possible as part of a pretty radical redesign of how tracing works...

Reply:
> I'd be happy to expand the message further if you think that case is common enough to provide debugging assistance for.

I'm less concerned that it's a particularly common case, and more concerned about inadvertently sending the message that passing an `&Id` to the span macro's `parent` field is the _only_ way this error might occur, which isn't really the case. Not a big deal either way, though.